File: google-maps-routes-nuevas/python/src/route_finder.py
#!/usr/bin/env python3
import requests
import json
import time

API_KEY = "AIzaSyCGbLfwvbxGV_bWvbtkHQrV8hHM7flIwMo"
ROUTES_URL = "https://routes.googleapis.com/directions/v2:computeRoutes"

def find_route(origin, destination, mode="DRIVE"):
    headers = {
        "Content-Type": "application/json",
        "X-Goog-Api-Key": API_KEY,
        "X-Goog-FieldMask": "routes.duration,routes.distanceMeters,routes.polyline.encodedPolyline,routes.legs.steps"
    }

    body = {
        "origin": {"address": origin},
        "destination": {"address": destination},
        "travelMode": mode,
        "routingPreference": "TRAFFIC_AWARE",
        "computeAlternativeRoutes": False,
        "languageCode": "es-ES",
        "units": "METRIC"
    }

    print(f"🔄 Calculando ruta usando Routes API...")
    response = requests.post(ROUTES_URL, headers=headers, json=body)

    if response.status_code == 200:
        data = response.json()
        if "routes" in data and data["routes"]:
            route = data["routes"][0]
            duration = int(route["duration"].replace("s", ""))
            hours = duration // 3600
            minutes = (duration % 3600) // 60
            duracion_str = f"{hours}h {minutes}m" if hours > 0 else f"{minutes}m"

            print(f"\n✅ Ruta encontrada!\n")
            print(f"📏 Distancia: {route['distanceMeters'] / 1000:.1f} km")
            print(f"⏱️ Duración: {duracion_str}")
            print(f"📝 Instrucciones (simplificadas):")

            for i, leg in enumerate(route.get("legs", [])):
                for j, step in enumerate(leg.get("steps", []), 1):
                    print(f"  {j}. {step.get('navigationInstruction', {}).get('instructions', 'Sigue recto')}")
        else:
            print("❌ No se encontraron rutas")
    else:
        print(f"❌ Error API: {response.status_code} - {response.text}")

if __name__ == "__main__":
    print("=== Routes API Finder (Python) ===")
    orig = input("Origen: ")
    dest = input("Destino: ")
    print("\nModos: 1=DRIVE, 2=WALK, 3=BIKE")
    mode_choice = input("Elige: ")
    mode = {"1": "DRIVE", "2": "WALK", "3": "BICYCLE"}.get(mode_choice, "DRIVE")
    find_route(orig, dest, mode)

File: google-maps-routes-nuevas/java/src/main/java/com/routes/RouteFinder.java
package com.routes;

import okhttp3.*;
import com.google.gson.*;
import java.io.IOException;
import java.util.Scanner;

public class RouteFinder {
    private static final String API_KEY = "AIzaSyCGbLfwvbxGV_bWvbtkHQrV8hHM7flIwMo";
    private static final String ROUTES_URL = "https://routes.googleapis.com/directions/v2:computeRoutes";
    private static final OkHttpClient client = new OkHttpClient();
    private static final Gson gson = new Gson();

    public static void main(String[] args) throws IOException {
        Scanner scanner = new Scanner(System.in);
        System.out.println("=== Routes API Finder (Java) ===");
        System.out.print("Origen: ");
        String origin = scanner.nextLine();
        System.out.print("Destino: ");
        String destination = scanner.nextLine();
        System.out.print("Modo (DRIVE/WALK/BICYCLE): ");
        String mode = scanner.nextLine().toUpperCase();

        findRoute(origin, destination, mode);
    }

    private static void findRoute(String origin, String destination, String mode) throws IOException {
        JsonObject body = new JsonObject();
        body.add("origin", createAddress(origin));
        body.add("destination", createAddress(destination));
        body.addProperty("travelMode", mode);
        body.addProperty("routingPreference", "TRAFFIC_AWARE");
        body.addProperty("languageCode", "es-ES");
        body.addProperty("units", "METRIC");

        Request request = new Request.Builder()
            .url(ROUTES_URL)
            .post(RequestBody.create(MediaType.parse("application/json"), gson.toJson(body)))
            .addHeader("Content-Type", "application/json")
            .addHeader("X-Goog-Api-Key", API_KEY)
            .addHeader("X-Goog-FieldMask", "routes.distanceMeters,routes.duration,routes.legs.steps")
            .build();

        try (Response response = client.newCall(request).execute()) {
            if (response.isSuccessful()) {
                JsonObject json = gson.fromJson(response.body().string(), JsonObject.class);
                JsonObject route = json.getAsJsonArray("routes").get(0).getAsJsonObject();
                double km = route.get("distanceMeters").getAsDouble() / 1000;
                int seconds = route.get("duration").getAsString().replace("s", "").isEmpty() ? 0 :
                              Integer.parseInt(route.get("duration").getAsString().replace("s", ""));
                System.out.printf("\n✅ Ruta encontrada!\n📏 %.1f km\n⏱️ %d min\n", km, seconds/60);
            } else {
                System.err.println("Error: " + response.body().string());
            }
        }
    }

    private static JsonObject createAddress(String address) {
        JsonObject obj = new JsonObject();
        obj.addProperty("address", address);
        return obj;
    }
}
